### PR TITLE
Added custom failure message to Mockito.verify. Issue 482

### PR DESCRIPTION
**Pull request by geoffschoeman**
Issue 482: passing custom failure message to Mockito.verify

### CODE REVIEW

**Comment by geoffschoeman** [comment]
@szczepiq, Any idea on when this will make it into master? I'll hold off on updating the @since annotations until you are ready to pull.

**Comment by szczepiq** [comment]
@geoffschoeman, this is a good change and we want to have it in mockito. Sorry I haven't merged it - I've been dealing with release automation and now I'm busy getting 2.0 final out. Perhaps @bric3 can merge it. Anyhow, we'll merge it at some point. Sorry for the delay and happy New Year!

**Comment by geoffschoeman** [comment]
@bric3, any chance that this will make it into the next build?

File: src/org/mockito/Mockito.java [hunk 1]
```diff
@@ -62,6 +62,7 @@
  *      <a href="#32">32. (new) Better generic support with deep stubs (Since 1.10.0)</a></h3>
  *      <a href="#32">33. (new) Mockito JUnit rule (Since 1.10.17)</a><br/>
  *      <a href="#34">34. (new) Switch <em>on</em> or <em>off</em> plugins (Since 1.10.15)</a><br/>
+ *      <a href="#35">35. (new)Custom verification failure message (Since 2.0.13)</a><br/>
```
Comment (bric3) [inline src/org/mockito/Mockito.java]: Space typo ;)

File: src/org/mockito/verification/VerificationMode.java [hunk 2]
```diff
@@ -33,4 +33,11 @@
     
     void verify(VerificationData data);
     
+    /**
+     * Description will be prepended to the assertion error if verification fails.
+     * @param description The custom failure message
+     * @return VerificationMode
+     * @since 2.0.3
+     */
+    VerificationMode description(String description);
```
Comment (bric3) [inline src/org/mockito/verification/VerificationMode.java]: I think a separate interface would be better. I'm leaning to something like `Descriptable`

But maybe a sub-interface like `DescriptableVerificationMode` could do it too.

File: test/org/mockito/internal/verification/VerificationWithDescriptionTest.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    @Mock 
+    private List mock;
+    
+    @Before
+    public void setUp() {
+        MockitoAnnotations.initMocks(this);
+    }
+    
+    @Test
+    public void test() {
```
Comment (bric3) [inline test/org/mockito/internal/verification/VerificationWithDescriptionTest.java]: if `test` could be more meaningful ;)

Comment (geoffschoeman) [inline test/org/mockito/internal/verification/VerificationWithDescriptionTest.java]: Good catch. Fixed

File: test/org/mockito/internal/verification/VerificationWithDescriptionTest.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+        String failureMessage = "Verification failed!";
+        try {
+            verify(mock, description(failureMessage)).clear();
+            fail("Should not have made it this far");
+            
+        } catch (MockitoAssertionError e) {
+            assert(e.getMessage().startsWith(failureMessage));
```
Comment (bric3) [inline test/org/mockito/internal/verification/VerificationWithDescriptionTest.java]: I'd prefer `assertTrue` or soemthing similar

File: test/org/mockito/internal/verification/DescriptionTest.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+import static org.mockito.MockitoAnnotations.initMocks;
+import org.mockito.exceptions.base.MockitoAssertionError;
+import org.mockito.internal.verification.api.VerificationData;
+import org.mockito.verification.VerificationMode;
+
+/**
+ *
+ * @author Geoff.Schoeman
+ */
+public class DescriptionTest {
```
Comment (bric3) [inline test/org/mockito/internal/verification/DescriptionTest.java]: unnecessary javadoc ;)

**Comment by bric3** [comment]
Also it may be worth to reflect this addition to `BDDMockito`

Before merging could you squash commits ?

File: src/org/mockito/internal/verification/AtLeast.java [hunk 6]
```diff
@@ -58,4 +58,12 @@ public void verifyInOrder(VerificationDataInOrder data) {
     public String toString() {
         return "Wanted invocations count: at least " + wantedCount;
     }
+    
+    /**
+     * {@inheritDoc}
+     */
+    @Override
```
Comment (bric3) [inline src/org/mockito/internal/verification/AtLeast.java]: unnecessary javadoc as well (all the `@inheritDoc`), it's eclipse that generates those stubs right ;)

**Comment by bric3** [comment]
OK, thanks for the updates :)

API may be tweaked in final 2.0.0. Note that a `Descriptable` interface may be extracted from this work, it'll help we we refactor hamcrest away from mockito

**Comment by bric3** [comment]
Oh and thanks for the PR as well

**Comment by geoffschoeman** [comment]
Thanks

**Pull request merged**